### DISCUSSION TITLE
Test helpers

### DISCUSSION THREAD

**Discussion by renchap** [janko/rodauth-rails#38] [Ideas]
I am continuing with some ideas from my Devise => Rodauth migration.

Devise provides some useful [controller](https://github.com/heartcombo/devise/blob/master/lib/devise/test/integration_helpers.rb) and [system/integration](https://github.com/heartcombo/devise/blob/master/lib/devise/test/controller_helpers.rb) helpers for testing and it would be great to have similar ones for Rodauth.

I implemented some for my system & integration tests using the informations in the readme, but those are making real auth requests and this is quite inefficient (my test suite is 30% slower compared to before as it needs to log in for most of the tests), it will fail every tests if auth breaks and you need to know the password for each user you are logging with.

Would it be possible to have some helpers to set `env["rodauth"].session_value` at the start of a test (`setup` block) and reset Rodauth's session in `teardown`?

You could then do

```ruby
# test_helper.rb
class ActionDispatch::IntegrationTest
  include ::RodauthTestHelpers
end

class ActionDispatch::SystemTestCase
  include ::RodauthTestHelpers
end

# my_test.rb

test "do something" do
  login user(:my_user))
  # logged in

  logout
  # no longer authed

  login user(:my_user), :admin
  # logged in as an admin
end
```

**Comment by janko** [comment]
The way that Devise implements these tests seems very hacky and complex, either relying on global state (system tests) or jumping through a lot of hoops to assign the right instance variables on the controller and Warden instances. I don't want rodauth-rails' test helpers to have this kind of complexity.

I was hoping it's possible to just set session values, but I'm not sure how to do it in integration tests. One would also need to remember to set all the necessary session values, as Rodauth doesn't work only with account ID.

Doing actual authentication is definitely the most robust, it's unfortunate that it seems to be adding so much overhead. I would be curious which parts of those auth requests are adding the most overhead.

**Reply by renchap** [reply]
In my app, the login screen is a React modal, so the browser needs to load the page, open the modal, log in, then reload the page. It could probably be faster with a simple form.
I looked at making a JSON request to the login url to initialise the session (this needs a valid CSRF token), but this does not work in system tests as Selenium does not allow have the browser do arbitrary requests.

I have not yet started to get my controller tests green (this is my next step) but I am wondering if this will work at all, I am not sure the middlewares are correctly invoked in controller testing.

Anyway I thought that this would be quite complex to get those helpers without doing hacky stuff like Devise, but I still feel like this would be quite useful :D

**Reply by renchap** [reply]
Forget what I said about controller tests, it works fine and a simple `post "/auth/login"` with the correct params logs in the user correctly.

**Reply by janko** [reply]
I've been exploring generating a Rails session cookie value from logged in session data (mainly `:account_id` set in the session data), with the goal of putting it into a cookie during browser tests, but it doesn't seem that's possible to do outside the request lifecycle.

I was also exploring if there is a way to tell the browser to immediately do a POST request to the login endpoint, to bypass any potentially expensive rendering of the login form, but Ferrum doesn't seem to support that.

An alternative could be to create a test login page with vanilla HTML (no JS or CSS), which system tests would use for fast login. But that would probably need to be implemented on the application side.

I don't have any better ideas at the moment.
